client/dcr: Add staking methods.

JoeGruffins opened the pull request:
part of #2264

This pr only adds capabilities to the asset and simnet testing framework.

Many of the changes are files copied from vspd and wallet. Hopefully the fee paying logic can be exported from wallet so we won't need to add here.

This will be updated.

martonp commented:
Would it make sense to have a separate VSP library in a new repo containing all the files you copied? And then vspd, dex, and other projects could just import it.

buck54321 commented:
> Would it make sense to have a separate VSP library in a new repo containing all the files you copied? And then vspd, dex, and other projects could just import it.

@jholdstock wanted to put it in the vspd repo anyway. Maybe @JoeGruffins could open up a pull request over there too. But we don't want to wait for them either. We'll delete this stuff once it's available.

chappjc commented:
If https://github.com/decred/dcrwallet/pull/2224 (or something like it happens), are we still left with a bunch of copied files?  I didn't see the discussion with @jholdstock about putting this stuff in the vspd repo.

JoeGruffins commented:
Asking on matrix if there were plans to move the rest of the dcrwallet/internal/vsp package to the decred/vspd repo.

Not sure about the files in `dex/testing/dcr/vspdtemplates` but the simnet vspd panics without them so I guess we need to copy to use vspd from the test folder location. I believe dexc also used to require files in a certain location like this but we build them into the executable at some point.

chappjc commented:
Echoing the diffs proposed on matrix:
- vspd: https://github.com/jholdstock/vspd/compare/dcr-deps...jholdstock:vspd:internal-vsp?expand=1
- dcrwallet: https://github.com/jholdstock/dcrwallet/compare/master...jholdstock:dcrwallet:internal-vsp?expand=1

buck54321 commented:
Could do [this](https://github.com/buck54321/dcrdex/compare/cb2868e...buck54321:dcrdex:use-jholdstock-for-now) while waiting for https://github.com/decred/vspd/pull/382.

buck54321 commented:
I was playing with this a little. Please consider [these changes](https://github.com/buck54321/dcrdex/compare/cb2868e...buck54321:dcrdex:dcr-vsp-ticket-detail-and-method-consolidation).

1) Consolidates pure getters into a single call
2) Adds extra ticket details
3) Prepares for pagination

JoeGruffins commented:
Made the buck change suggestions. For some reason go mod tidy does not error when I run it but does with the testing script. `invalid version: unknown revision 000000000000`

buck54321 commented on client/asset/dcr/spv_test.go:
```diff
@@ -27,8 +28,13 @@ import (
+}
+
+type tDcrWallet struct {
+	nb // Using wallet.NetworkBackend causes a naming conflict with the
```
Can embed `spvSyncer` instead.

JoeGruffins replied:
Did this.

buck54321 reviewed (changes requested):
Do we need to do something with `(*AutoClient).ProcessUnprocessedTickets`?

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -4524,6 +4569,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+}
+
+func (dcr *ExchangeWallet) StakeStatus() (*asset.TicketStakingStatus, error) {
+	if !dcr.connected.Load() {
+		return nil, errors.New("not connected, login first")
+	}
```
My memory is fuzzy, but these will panic if the dcr.ctx is nil, which is so before the first connect.

JoeGruffins replied:
Yeah, I added this to any methods that use `dcr.ctx` to avoid using a nil context and panic inside the rpcclient. Should I handle this differently?

buck54321 commented:
Can get rid of the replace and rebase now, I think. Probably good to pull out of draft too.

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -803,6 +825,11 @@ func unconnectedWallet(cfg *asset.WalletConfig, dcrCfg *walletConfig, chainParam
 
+	dir := filepath.Join(cfg.DataDir, chainParams.Name)
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
```
Added context.

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -4524,6 +4569,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+}
+
+// PurchaseTickets purchases n amout of tickets. Ability to purchase should be
+// checked with CanPurchaseTickets. Part of the asset.TicketBuyer interface.
```
Yeah. The original plan was to check the wallet config before purchasing over rpc, and we would deny if not using a vsp. But, giving up on that since it seems we cannot get the config. User will just have to know what they are doing. Anyway, removing the reference to the removed method.

JoeGruffins commented on client/asset/interface.go:
```diff
@@ -797,6 +808,72 @@ type PeerManager interface {
+	// CanSetVSP first.
+	SetVSP(addr string) error
+	// PurchaseTickets purchases n amout of tickets. Ability to purchase
+	// should be checked with CanPurchaseTickets. Returns the purchased
```
Removed comment.

JoeGruffins commented on client/asset/dcr/spv.go:
```diff
@@ -823,6 +836,272 @@ func (w *spvWallet) AddressPrivKey(ctx context.Context, addr stdaddr.Address) (*
+		_, endBlockNum = w.MainChainTip(ctx)
+	}
+	startBlockNum := fromHeight
+	if startBlockNum > 0 {
```
Yeah... This looks like a mistake. Changing to `startBlockNum == 0`

JoeGruffins commented on client/asset/dcr/spv.go:
```diff
@@ -823,6 +836,272 @@ func (w *spvWallet) AddressPrivKey(ctx context.Context, addr stdaddr.Address) (*
+			w.log.Warnf("unable to get VSP associated with ticket %s: %v", hash, err)
+			return nil
+		}
+		info, err := vspInfo(vspHost)
```
Good idea. A map local to the function good enough?

JoeGruffins commented:
buck changes https://github.com/decred/dcrdex/compare/0eaf8a3d498cc05e4d7646cba9e89e1851ea757d..3f91df5d9253ec15136656a877e44e29ded1b22d

JoeGruffins commented:
> Do we need to do something with (*AutoClient).ProcessUnprocessedTickets?

Maybe. Client could do this if wallet was reset from seed I think. Is it ok to make a todo for now? Because of automatic revocations now, its not a HUGE deal even if the vsp is not set up. Also, if the fee was paid, and the vspd is ok, the wallet doesnt need to worry about it, really. The vsp will vote and the wallet will get the funds back at some point.

JoeGruffins commented:
Made a TODO for managing unmanaged tickets https://github.com/decred/dcrdex/compare/be02793d87ec38b68e79012b4a1b0c047643f761..82014c37e3f680b799345fd1f08bdb429089594f

JoeGruffins commented:
So, copying the vspd client logic for now but using dcrwallet/v2. Hopefully can be removed sooner or later, but this seems like the only way to continue immediately.

buck54321 commented:
vspd cleaned up their go.mod and chappjc is working through the dcrwallet v3 upgrade now.

chappjc commented on client/asset/dcrspvclient/client.go:
```diff
@@ -0,0 +1,252 @@
+// decred/vspd when possible. That will probably be when the replaces there are
+// gone and this repo is using dcrwallet/v3.
+
+package dcrspvclient
```
I think you can get in dcrwallet/v3 now.  Hopefully code duplication is mostly gone now.

Also, I'm scratching my head over dcr**spv**client.  Should that have been dcr**vsp**client?

JoeGruffins replied:
> Should that have been dcrvspclient?

yes

JoeGruffins commented:
Using github.com/decred/vspd/client/v2 now https://github.com/decred/dcrdex/compare/a374b1632ca4126184d094b37d1e5907b0d4a879..7d192d63c92c9483324b634a0d694c0dede19a6c

chappjc commented:
Nice work l. 💪

JoeGruffins commented on client/webserver/site/package-lock.json:
```diff
@@ -1,12 +1,12 @@
 {
   "name": "dexclient",
-  "version": "1.1.0",
+  "version": "1.2.0",
   "lockfileVersion": 2,
   "requires": true,
   "packages": {
     "": {
       "name": "dexclient",
-      "version": "1.1.0",
+      "version": "1.2.0",
       "license": "Blue Oak 1.0.0",
```
This thing checked in here too. Reverting...

JoeGruffins replied:
Reverted. https://github.com/decred/dcrdex/compare/7d192d63c92c9483324b634a0d694c0dede19a6c..ce32862ef975be44d1370b19382bc543dbc814bb

This is set to `1.2.0` whenever I run `npm install`

chappjc replied:
> Reverted. https://github.com/decred/dcrdex/compare/7d192d63c92c9483324b634a0d694c0dede19a6c..ce32862ef975be44d1370b19382bc543dbc814bb
> 
> This is set to `1.2.0` whenever I run `npm install`

You should only run `npm clean-install` unless you are trying to update a dependency to something that is beyond what's specified in package-lock.json.

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -4524,6 +4569,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+	return nil
+}
+
+// PurchaseTickets purchases n amout of tickets. Ability to purchase should be
```
Fixing

chappjc commented on client/cmd/dexc-desktop/go.mod:
```diff
@@ -5,7 +5,7 @@ go 1.18
 
 require (
-	decred.org/dcrdex v0.6.1
+	decred.org/dcrdex v0.6.0
```
looks like an unintentional rollback

JoeGruffins replied:
Fixed! I guess this value doesn't really matter with the replace?

chappjc replied:
Sorry to dwell on this, but all the indirect depends resulting from this are still changed.  Best just to restore these files from master e.g. `git restore --source=master go.mod`

JoeGruffins replied:
Ok! Will add that command to my knowledge.

chappjc commented on dex/testing/dcr/vspdtemplates/header.html:
```diff
@@ -0,0 +1,76 @@
+
+        <title>Decred VSP - {{ .WebApiCfg.Designation }}</title>
+
+        <link rel="stylesheet" href="/public/css/vendor/bootstrap-4.5.0.min.css" />
+        <link rel="stylesheet" href="/public/css/vspd.css?v={{ .WebApiCfg.VspdVersion }}" />
+        <!-- fonts.css should be last to ensure dcr fonts take precedence. -->
+        <link rel="stylesheet" href="/public/css/fonts.css?v={{ .WebApiCfg.VspdVersion }}" />
+
+        <!--  Custom favicon  -->
+        <!-- Apple PWA -->
+        <link rel="apple-touch-icon" sizes="57x57"   href="/public/images/favicon/apple-touch-icon-57x57.png?v=gT6Mc">
```
Looks like we don't intend for the web interface for vspd to actually work?  We're not checking in the stuff in /public (and I don't want to), so are these html templates just in the repository so vspd doesn't choke on startup?  If so, we can just make stubs for all these html files that define mostly-empty templates that execute without error.

chappjc replied:
Or should we have git clone of vspd be part of create-vpsd.sh since vspd doesn't (yet) embed these resources.

JoeGruffins replied:
> Or should we have git clone of vspd be part of create-vpsd.sh since vspd doesn't (yet) embed these resources.

This is the better idea. Will do.

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -825,6 +847,11 @@ func unconnectedWallet(cfg *asset.WalletConfig, dcrCfg *walletConfig, chainParam
 	}
 
+	dir := filepath.Join(cfg.DataDir, chainParams.Name)
+	if err := os.MkdirAll(dir, 0755); err != nil {
```
Although we use `0755` elsewhere in the repo, I'm thinking we should set this to `0744`. Only users should r,w,x others should only r.

chappjc replied:
`x` is the permission bit to list a folders contents. `r` alone does not do that.  If you deny execute for a folder, you might as well deny read as well.

JoeGruffins replied:
Leaving I guess? I don't know if other values have some unintended effects.

I think I added this so there is a place for for the vsp file, but I think it is made at some point after with the 0755 permissions if not here....

ukane-philemon replied:
> Leaving I guess? 

Yes, we can keep `0755`. @chappjc agrees it's better.

We also use `0744` elsewhere, but maybe it does not apply.

chappjc replied:
> We also use `0744` elsewhere, but maybe it does not apply.

That sounds like an error and/or copy pasta that has spread.  There are indeed many such lines, and git blame says that myself and others have done this, hence it smells like copy pasta.

Anyone else have an opinion about this?  0744 makes little sense to me.

buck54321 replied:
755

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -836,6 +863,22 @@ func unconnectedWallet(cfg *asset.WalletConfig, dcrCfg *walletConfig, chainParam
+		walletType:          cfg.Type,
+	}
+
+	if b, err := os.ReadFile(filepath.Join(dir, vspFileName)); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, fmt.Errorf("unable to read vsp file: %v", err)
+		}
+	} else {
+		var v vsp
+		err = json.Unmarshal(b, &v)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal vsp file: %v", err)
+		}
+		w.vspV.Store(&v)
 	}
```
We could avoid the nested `if`?

```
	if b, err := os.ReadFile(filepath.Join(dir, vspFileName)); err == nil {
		var v vsp
		err = json.Unmarshal(b, &v)
		if err != nil {
			return nil, fmt.Errorf("unable to unmarshal vsp file: %v", err)
		}
		w.vspV.Store(&v)
	} else if !errors.Is(err, os.ErrNotExist) {
		return nil, fmt.Errorf("unable to read vsp file: %v", err)
	}```

JoeGruffins replied:
Ok, changed.

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -900,6 +943,8 @@ func (dcr *ExchangeWallet) Connect(ctx context.Context) (*sync.WaitGroup, error)
 	defer func() {
 		if !success {
 			dcr.wallet.Disconnect()
+		} else {
+			dcr.connected.Store(true)
```
I'm thinking we can move this to L970 below. Where we set success to `true`.

JoeGruffins replied:
Ok.

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -4546,6 +4591,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
 	return finalFee, isValidAddress, nil
 }
 
+func (dcr *ExchangeWallet) isInternal() bool {
```
`isSPV`(or anything with spv in it?) relates better, I think.

JoeGruffins replied:
Changed to `isSPV`

ukane-philemon reviewed (changes requested):
utAck. Good work @JoeGruffins.

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -555,6 +561,14 @@ type mempoolRedeem struct {
+// vsp holds info needed for purchasing tickes from a vsp. pub is from the vsp
+// and is used for verifying communications.
+type vsp struct {
+	Url           string  `json:"url"`
```
Changed.

JoeGruffins commented:
Just clone vspd for web files and fix some names https://github.com/decred/dcrdex/compare/83e734a95df1af93f6a5c3b1c4eb345bd816b10f..2f64728b70a0292a544718cd0b83d29c08eaa914

Fix dexc-desktop version in go.mod https://github.com/decred/dcrdex/compare/2f64728b70a0292a544718cd0b83d29c08eaa914..e35ffb5416101fe991099a47bfe0aabfbefd1039

JoeGruffins commented:
`git restore --source=master go.mod` https://github.com/decred/dcrdex/compare/e35ffb5416101fe991099a47bfe0aabfbefd1039..577da2e864436624888cd50a34fbf7b0d7c76ed5

JoeGruffins commented:
go mod tidy... https://github.com/decred/dcrdex/compare/577da2e864436624888cd50a34fbf7b0d7c76ed5..29d50b0a7e546a0a63499f5b889df74e8448c05d

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -5117,6 +5159,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+	return &info, nil
+}
+
+// SetVSP sets the VSP provider. Ability to set should be checked with CanSetVSP
+// first. Part of the asset.TicketBuyer interface.
```
I can't find `CanSetVSP` anywhere in the code. Yet to implement?

JoeGruffins replied:
The ability to detect if we can set was added to the returns from `StakeStatus()` `IsRPC` Will remove from comment.

JoeGruffins replied:
Changed.

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -5117,6 +5159,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+
+// SetVSP sets the VSP provider. Ability to set should be checked with CanSetVSP
+// first. Part of the asset.TicketBuyer interface.
+func (dcr *ExchangeWallet) SetVSP(url string) error {
```
It seems `SetVSP` and other `TicketBuyer` methods are not consumed anywhere, except in `simnet_test.go`, this means consumption of these new methods/features would be in a follow-up PR, no?

JoeGruffins replied:
Yes. With the rpcserver in #2316 and webserver not done yet.

ukane-philemon commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+	err = ew.SetVSP(vspAddr)
+	if isInternal {
+		if err != nil {
+			t.Fatalf("unexected error setting spv for internal wallet: %v", err)
```
nit: `unexected`

JoeGruffins replied:
fixed

ukane-philemon commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+		}
+	} else {
+		if err == nil {
+			t.Fatal("exected error setting spv for external wallet")
```
nit: `unexected`

ukane-philemon commented on client/asset/dcr/dcr.go:
```diff
@@ -5117,6 +5159,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+	v := dcr.vspV.Load()
+	if v == nil {
+		return nil, errors.New("no vsp set")
+	}
+	return dcr.wallet.PurchaseTickets(dcr.ctx, n, v.(*vsp).URL, v.(*vsp).PubKey)
```
nit: Maybe we should convert once? 
```
vspInfo := v.(*vsp)
....., vspInfo.URL, vspInfo.PubKey)
```

JoeGruffins replied:
OK

ukane-philemon commented on client/asset/dcr/wallet.go:
```diff
@@ -143,7 +145,19 @@ type Wallet interface {
-
+	// StakeDiff gets the stake difficulty.
+	StakeDiff(ctx context.Context) (dcrutil.Amount, error)
+	// PurchaseTickets purchases n tickets. vsp arguments only needed for
+	// internal wallets.
+	PurchaseTickets(ctx context.Context, n int, vspHost, vspPubKey string) ([]string, error)
```
Please can we specify the exact arguments that are not needed for `RPC` wallets? I see only `vspHost` and `vspPubKey` are ignored and all other arguments are provided. `vsp arguments` might not ring a bell the first time.

JoeGruffins replied:
`vspHost and vspPubKey only needed for internal wallets.` okay?

buck54321 commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+	tLogger.Infof("Purchased the following tickets: %v", tickets)
+
+	// Test SetVotingPreferences.
+	if err := ew.SetVotingPreferences(nil, nil, nil); err != nil {
+		t.Fatalf("error setting voting preferences: %v", err)
+	}
+
+	// Test StakeStatus again.
+	ss, err = ew.StakeStatus()
+	if err != nil {
+		t.Fatalf("unable to get stake status: %v", err)
+	}
```
We should set the voting preferences for whatever is the latest deployment version, and then check that the values claimed in the `StakeStatus` match what we just set.

JoeGruffins replied:
Ah yes, I dont think this was possible without a lot of work until recently.

JoeGruffins replied:
Added but not sure how to allow for new deployments...

Also setting the `tspendPolicy` doesn't seem to be remembered? Maybe it needs an active tSpend? Or I could have implemented it wrong or it could be a bug.

buck54321 replied:
Some suggestions https://github.com/buck54321/dcrdex/compare/c3b9e76...buck54321:dcrdex:staking-methods-harness-tests

buck54321 commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+	}
+
+	// Test StakeStatus again.
+	ss, err = ew.StakeStatus()
+	if err != nil {
+		t.Fatalf("unable to get stake status: %v", err)
+	}
```
Wanna check that there's 3 tickets here?

JoeGruffins replied:
If this is the rpcwallet there will be a ton of tickets.

buck54321 commented on client/asset/dcr/simnet_test.go:
```diff
@@ -58,18 +63,22 @@ func mineAlpha() error {
+		if err != nil {
+			t.Fatalf("error reading config options: %v", err)
+		}
+		settings["account"] = "default"
```
Can actually drop this now. We're doing [this](https://github.com/decred/dcrdex/blob/4ab15ac46abf273caa55f898360c07b0a4a8a0c4/client/asset/dcr/dcr.go#L784-L787).

JoeGruffins replied:
Removed `settings["account"] = "default"`

buck54321 commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+	defer rig.close(t)
+	tLogger.Info("Testing ticket methods with the alpha rig.")
+	testTickets(t, false, rig.alpha())
+	// TODO: beta wallet is spv but has no vps in its config. Add it and
```
vps -> vsp

JoeGruffins replied:
fixed

buck54321 reviewed (changes requested):
Mostly just some nits for now. Testing well.

JoeGruffins commented on dex/testing/dcr/create-vspd.sh:
```diff
@@ -0,0 +1,51 @@
+PORT=$2
+FEE_XPUB=$3
+
+VSPD_DIR="${NODES_ROOT}/vspd"
+
+git clone -b client/v2.1.0 https://github.com/decred/vspd ${VSPD_DIR}
```
Isn't the dir in `$HOME/dextest` ? I think the whole dcr folder is deleted every harness start. https://github.com/decred/dcrdex/blob/95e0e5806e298624d7c4be19f6a985a37dbc46ef/dex/testing/dcr/harness.sh#L46-L48

buck54321 replied:
Oh. Whoops.

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -609,6 +625,11 @@ type ExchangeWallet struct {
 	mempoolRedeems    map[[32]byte]*mempoolRedeem // keyed by secret hash
+
+	vspV    atomic.Value // *vsp
+	vspInfo func(url string) (*vspdjson.VspInfoResponse, error)
```
I thought I was doing this for testing but I don't see any tests now...

JoeGruffins replied:
Removed.

JoeGruffins commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+	err = ew.SetVSP(vspAddr)
+	if isInternal {
+		if err != nil {
+			t.Fatalf("unexected error setting spv for internal wallet: %v", err)
```
Sorry, think I did this all over. fixed.

JoeGruffins commented on client/asset/dcr/simnet_test.go:
```diff
@@ -531,3 +577,62 @@ func runTest(t *testing.T, splitTx bool) {
+		}
+	} else {
+		if err == nil {
+			t.Fatal("exected error setting spv for external wallet")
```
fixed

JoeGruffins commented on dex/testing/dcr/harness.sh:
```diff
@@ -297,23 +309,39 @@ tmux send-keys -t $SESSION:0 "./mine-alpha 2${WAIT}" C-m\; wait-for donedcr
+echo "Creating simnet vspd wallet"
+USE_SPV="0"
+ENABLE_VOTING="1"
+MANUAL_TICKETS="1"
+"${HARNESS_DIR}/create-wallet.sh" "$SESSION:7" "vspdwallet" ${VSPD_WALLET_SEED} \
+${VSPD_WALLET_RPC_PORT} ${USE_SPV} ${ENABLE_VOTING} "_" ${MANUAL_TICKETS}
```
Oh I don't really know, just following the guide.

`dcrwallet is also required to have the manual tickets option (--manualtickets) enabled which disables dcrwallet adding tickets arriving over the network.`

https://github.com/decred/vspd/blob/master/docs/deployment.md#voting-servers

JoeGruffins commented on dex/testing/dcr/harness.sh:
```diff
@@ -297,23 +309,39 @@ tmux send-keys -t $SESSION:0 "./mine-alpha 2${WAIT}" C-m\; wait-for donedcr
+"${HARNESS_DIR}/create-wallet.sh" "$SESSION:7" "vspdwallet" ${VSPD_WALLET_SEED} \
+${VSPD_WALLET_RPC_PORT} ${USE_SPV} ${ENABLE_VOTING} "_" ${MANUAL_TICKETS}
+
+#sleep 15
+sleep 3
```
deleted...

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -5117,6 +5159,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+	return nil
+}
+
+// PurchaseTickets purchases n amount of tickets. Part of the asset.TicketBuyer
```
fixed

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -5117,6 +5159,123 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
+	return dcr.wallet.PurchaseTickets(dcr.ctx, n, v.(*vsp).URL, v.(*vsp).PubKey)
+}
+
+// SetVotingPreferences sets default voting settings for all active tickets and
+// future tickets. Nil maps can be provided for no change. Part of the
+// asset.TicketBuyer interface.
```
fixed

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -838,6 +865,20 @@ func unconnectedWallet(cfg *asset.WalletConfig, dcrCfg *walletConfig, chainParam
+		walletType:          cfg.Type,
+	}
+
+	if b, err := os.ReadFile(filepath.Join(dir, vspFileName)); err == nil {
+		var v vsp
+		err = json.Unmarshal(b, &v)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal vsp file: %v", err)
+		}
+		w.vspV.Store(&v)
+	} else if !errors.Is(err, os.ErrNotExist) {
```
I guess could do that.

JoeGruffins replied:
Ok doing that.

buck54321 replied:
I prefer the file-based way with exported `SetVSP` being the only way to set it. As it is now, if the user runs `(TicketBuyer).SetVSP`, their changes would not persist. I also want to leave **core** out of `TicketBuyer` business as much as possible, e.g. storing ticket-related configuration.

I could maybe see offering both options, but much prefer the old way.

JoeGruffins replied:
Reverted.

JoeGruffins commented on client/asset/interface.go:
```diff
@@ -116,6 +118,12 @@ func (wt WalletTrait) IsTokenApprover() bool {
 	return wt&WalletTraitTokenApprover != 0
 }
 
+// IsTicketBuyer tests if the WalletTrait has the WalletTraitDCRStaker bit set,
```
fixed

JoeGruffins commented:
Addressing reviews https://github.com/decred/dcrdex/compare/467337e1993823a1ffcac08d0e966282cc789548..c3b9e76eae170af3e91a6573d20c870ad2b36857

Notable changes are removing the vsp file and just getting the fee and pubkey on wallet start and adding some test parameters.

JoeGruffins commented:
ty @buck54321 https://github.com/decred/dcrdex/compare/877601896fc96501c2ba2ac01b837d234d21187b..bd79ca33db163d89bd2ae4e1080afb5109c2e1b8

Also using the separate file to save again.

buck54321 reviewed (approved):
Good stuff. I'll follow up with a couple of tweaks.

itswisdomagain commented on client/asset/dcr/dcr.go:
```diff
@@ -5134,6 +5174,125 @@ func (dcr *ExchangeWallet) EstimateSendTxFee(address string, sendAmount, feeRate
 }
 
+func (dcr *ExchangeWallet) isSPV() bool {
+	return dcr.walletType == walletTypeSPV
```
This reminds me. There's a `dcr.wallet.SpvMode()` which may be true even for a wallet connected via rpc. Put another way, isSPV could be true even if walletType isn't walletTypeSPV.
